File: backend/api/smart_preload.py
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from sqlalchemy.orm import Session
from typing import List, Optional, Dict, Any
from pydantic import BaseModel
import logging

from models.database import get_db
from services.smart_preload_service import smart_preload_service, PreloadPriority
from utils.middleware import require_admin, get_current_user
from models.user import User

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/smart-preload", tags=["smart-preload"])

# Pydantic models
class SmartPreloadRequest(BaseModel):
    playlist_id: int
    current_image_index: int = 0
    slide_duration: int = 5
    formats: List[str] = ["webp", "avif"]

class PreloadStatusResponse(BaseModel):
    image_id: Optional[int] = None
    status: str
    priority: Optional[str] = None
    retry_count: Optional[int] = None
    error: Optional[str] = None
    completed_at: Optional[str] = None
    processing_time: Optional[float] = None

class OverallStatusResponse(BaseModel):
    worker_running: bool
    queue_size: int
    active_tasks: int
    completed_tasks: int
    failed_tasks: int
    memory_stats: Dict[str, Any]
    stats: Dict[str, Any]
    max_concurrent_tasks: int
    memory_threshold_percent: float

class SlideshowTimingUpdate(BaseModel):
    playlist_id: int
    current_index: int
    slide_duration: int

@router.post("/start")
async def start_smart_preload_worker():
    """Start the smart preload worker"""
    try:
        await smart_preload_service.start_worker()
        return {"message": "Smart preload worker started successfully"}
    except Exception as e:
        logger.error(f"Failed to start smart preload worker: {e}")
        raise HTTPException(status_code=500, detail="Failed to start worker")

@router.post("/stop")
async def stop_smart_preload_worker():
    """Stop the smart preload worker"""
    try:
        await smart_preload_service.stop_worker()
        return {"message": "Smart preload worker stopped successfully"}
    except Exception as e:
        logger.error(f"Failed to stop smart preload worker: {e}")
        raise HTTPException(status_code=500, detail="Failed to stop worker")

@router.post("/preload-playlist")
async def preload_playlist_smart(
    request: SmartPreloadRequest,
    background_tasks: BackgroundTasks
):
    """Start smart preloading for a playlist"""
    try:
        result = await smart_preload_service.preload_playlist_smart(
            playlist_id=request.playlist_id,
            current_image_index=request.current_image_index,
            slide_duration=request.slide_duration,
            formats=request.formats
        )
        
        return result
        
    except Exception as e:
        logger.error(f"Failed to start smart preload: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to start preload: {str(e)}")

@router.get("/status", response_model=OverallStatusResponse)
async def get_overall_status():
    """Get overall smart preload status"""
    try:
        status = smart_preload_service.get_preload_status()
        return OverallStatusResponse(**status)
    except Exception as e:
        logger.error(f"Failed to get overall status: {e}")
        raise HTTPException(status_code=500, detail="Failed to get status")

@router.get("/status/{image_id}", response_model=PreloadStatusResponse)
async def get_image_status(image_id: int):
    """Get preload status for a specific image"""
    try:
        status = smart_preload_service.get_preload_status(image_id)
        return PreloadStatusResponse(**status)
    except Exception as e:
        logger.error(f"Failed to get image status: {e}")
        raise HTTPException(status_code=500, detail="Failed to get image status")

@router.post("/update-timing")
async def update_slideshow_timing(request: SlideshowTimingUpdate):
    """Update slideshow timing for optimization"""
    try:
        smart_preload_service.update_slideshow_timing(
            playlist_id=request.playlist_id,
            current_index=request.current_index,
            slide_duration=request.slide_duration
        )
        
        return {
            "message": "Slideshow timing updated successfully",
            "playlist_id": request.playlist_id,
            "current_index": request.current_index,
            "slide_duration": request.slide_duration
        }
        
    except Exception as e:
        logger.error(f"Failed to update slideshow timing: {e}")
        raise HTTPException(status_code=500, detail="Failed to update timing")

@router.get("/timing/{playlist_id}")
async def get_slideshow_timing(playlist_id: int):
    """Get current slideshow timing for a playlist"""
    try:
        timing = smart_preload_service.get_slideshow_timing(playlist_id)
        if timing:
            return timing
        else:
            return {
                "playlist_id": playlist_id,
                "message": "No timing data found for this playlist"
            }
    except Exception as e:
        logger.error(f"Failed to get slideshow timing: {e}")
        raise HTTPException(status_code=500, detail="Failed to get timing")

@router.post("/clear-completed")
async def clear_completed_tasks(older_than_hours: int = 1):
    """Clear completed tasks older than specified hours"""
    try:
        cleared_count = smart_preload_service.clear_completed_tasks(older_than_hours)
        return {
            "message": f"Cleared {cleared_count} completed tasks",
            "cleared_count": cleared_count,
            "older_than_hours": older_than_hours
        }
    except Exception as e:
        logger.error(f"Failed to clear completed tasks: {e}")
        raise HTTPException(status_code=500, detail="Failed to clear tasks")

@router.get("/memory-stats")
async def get_memory_stats():
    """Get detailed memory statistics"""
    try:
        memory_stats = smart_preload_service._get_memory_stats()
        return {
            "total_memory_mb": memory_stats.total_memory // (1024 * 1024),
            "available_memory_mb": memory_stats.available_memory // (1024 * 1024),
            "used_memory_mb": memory_stats.used_memory // (1024 * 1024),
            "used_memory_percent": memory_stats.memory_percent,
            "cache_memory_mb": memory_stats.cache_memory // (1024 * 1024),
            "memory_pressure": memory_stats.memory_percent > smart_preload_service.memory_threshold_percent
        }
    except Exception as e:
        logger.error(f"Failed to get memory stats: {e}")
        raise HTTPException(status_code=500, detail="Failed to get memory stats")

@router.post("/configure")
async def configure_preload_service(
    max_concurrent_tasks: Optional[int] = None,
    memory_threshold_percent: Optional[float] = None,
    max_cache_size_mb: Optional[int] = None,
    current_user: User = Depends(require_admin)
):
    """Configure smart preload service settings (admin only)"""
    try:
        if max_concurrent_tasks is not None:
            smart_preload_service.max_concurrent_tasks = max_concurrent_tasks
        
        if memory_threshold_percent is not None:
            smart_preload_service.memory_threshold_percent = memory_threshold_percent
        
        if max_cache_size_mb is not None:
            smart_preload_service.max_cache_size_mb = max_cache_size_mb
        
        return {
            "message": "Configuration updated successfully",
            "max_concurrent_tasks": smart_preload_service.max_concurrent_tasks,
            "memory_threshold_percent": smart_preload_service.memory_threshold_percent,
            "max_cache_size_mb": smart_preload_service.max_cache_size_mb
        }
        
    except Exception as e:
        logger.error(f"Failed to configure preload service: {e}")
        raise HTTPException(status_code=500, detail="Failed to update configuration")

@router.get("/stats")
async def get_preload_stats():
    """Get preload statistics"""
    try:
        return {
            "preload_stats": smart_preload_service.preload_stats,
            "current_config": {
                "max_concurrent_tasks": smart_preload_service.max_concurrent_tasks,
                "memory_threshold_percent": smart_preload_service.memory_threshold_percent,
                "max_cache_size_mb": smart_preload_service.max_cache_size_mb
            }
        }
    except Exception as e:
        logger.error(f"Failed to get preload stats: {e}")
        raise HTTPException(status_code=500, detail="Failed to get stats")

@router.post("/emergency-stop")
async def emergency_stop_preload():
    """Emergency stop all preload operations"""
    try:
        # Stop the worker
        await smart_preload_service.stop_worker()
        
        # Clear all queues and tasks
        while not smart_preload_service.preload_queue.empty():
            try:
                smart_preload_service.preload_queue.get_nowait()
            except:
                break
        
        # Clear active tasks
        smart_preload_service.active_tasks.clear()
        
        return {
            "message": "Emergency stop completed",
            "active_tasks_cleared": len(smart_preload_service.active_tasks),
            "queue_cleared": True
        }
        
    except Exception as e:
        logger.error(f"Failed to emergency stop: {e}")
        raise HTTPException(status_code=500, detail="Failed to emergency stop")

File: frontend/src/services/smartPreload.ts
import { apiService } from './api';

export interface PreloadRequest {
  playlist_id: number;
  current_image_index?: number;
  slide_duration?: number;
  formats?: string[];
}

export interface PreloadStatus {
  image_id?: number;
  status: string;
  priority?: string;
  retry_count?: number;
  error?: string;
  completed_at?: string;
  processing_time?: number;
}

export interface OverallStatus {
  worker_running: boolean;
  queue_size: number;
  active_tasks: number;
  completed_tasks: number;
  failed_tasks: number;
  memory_stats: {
    total_memory_mb: number;
    available_memory_mb: number;
    used_memory_percent: number;
    cache_size_mb: number;
  };
  stats: {
    total_preloaded: number;
    total_failed: number;
    total_retries: number;
    cache_hits: number;
    cache_misses: number;
    memory_pressure_events: number;
  };
  max_concurrent_tasks: number;
  memory_threshold_percent: number;
}

export interface MemoryStats {
  total_memory_mb: number;
  available_memory_mb: number;
  used_memory_mb: number;
  used_memory_percent: number;
  cache_memory_mb: number;
  memory_pressure: boolean;
}

export interface SlideshowTiming {
  playlist_id: number;
  current_index: number;
  slide_duration: number;
  last_updated: string;
}

export class SmartPreloadService {
  private statusUpdateInterval: NodeJS.Timeout | null = null;
  private onStatusUpdate: ((status: OverallStatus) => void) | null = null;
  private onMemoryUpdate: ((memory: MemoryStats) => void) | null = null;

  /**
   * Start the smart preload worker
   */
  async startWorker(): Promise<void> {
    try {
      const response = await apiService.api.post('/smart-preload/start');
      if (response.status !== 200) {
        throw new Error('Failed to start smart preload worker');
      }
    } catch (error) {
      console.error('Error starting smart preload worker:', error);
      throw error;
    }
  }

  /**
   * Stop the smart preload worker
   */
  async stopWorker(): Promise<void> {
    try {
      const response = await apiService.api.post('/smart-preload/stop');
      if (response.status !== 200) {
        throw new Error('Failed to stop smart preload worker');
      }
    } catch (error) {
      console.error('Error stopping smart preload worker:', error);
      throw error;
    }
  }

  /**
   * Start smart preloading for a playlist
   */
  async preloadPlaylist(request: PreloadRequest): Promise<any> {
    try {
      const response = await apiService.api.post('/smart-preload/preload-playlist', request);
      return response.data;
    } catch (error) {
      console.error('Error starting smart preload:', error);
      throw error;
    }
  }

  /**
   * Get overall preload status
   */
  async getOverallStatus(): Promise<OverallStatus> {
    try {
      const response = await apiService.api.get('/smart-preload/status');
      return response.data;
    } catch (error) {
      console.error('Error getting overall status:', error);
      throw error;
    }
  }

  /**
   * Get preload status for a specific image
   */
  async getImageStatus(imageId: number): Promise<PreloadStatus> {
    try {
      const response = await apiService.api.get(`/smart-preload/status/${imageId}`);
      return response.data;
    } catch (error) {
      console.error('Error getting image status:', error);
      throw error;
    }
  }

  /**
   * Update slideshow timing for optimization
   */
  async updateSlideshowTiming(playlistId: number, currentIndex: number, slideDuration: number): Promise<void> {
    try {
      await apiService.api.post('/smart-preload/update-timing', {
        playlist_id: playlistId,
        current_index: currentIndex,
        slide_duration: slideDuration
      });
    } catch (error) {
      console.error('Error updating slideshow timing:', error);
      throw error;
    }
  }

  /**
   * Get slideshow timing for a playlist
   */
  async getSlideshowTiming(playlistId: number): Promise<SlideshowTiming | null> {
    try {
      const response = await apiService.api.get(`/smart-preload/timing/${playlistId}`);
      return response.data;
    } catch (error) {
      console.error('Error getting slideshow timing:', error);
      return null;
    }
  }

  /**
   * Clear completed tasks older than specified hours
   */
  async clearCompletedTasks(olderThanHours: number = 1): Promise<{ cleared_count: number }> {
    try {
      const response = await apiService.api.post('/smart-preload/clear-completed', null, {
        params: { older_than_hours: olderThanHours }
      });
      return response.data;
    } catch (error) {
      console.error('Error clearing completed tasks:', error);
      throw error;
    }
  }

  /**
   * Get detailed memory statistics
   */
  async getMemoryStats(): Promise<MemoryStats> {
    try {
      const response = await apiService.api.get('/smart-preload/memory-stats');
      return response.data;
    } catch (error) {
      console.error('Error getting memory stats:', error);
      throw error;
    }
  }

  /**
   * Configure preload service settings (admin only)
   */
  async configureService(config: {
    max_concurrent_tasks?: number;
    memory_threshold_percent?: number;
    max_cache_size_mb?: number;
  }): Promise<void> {
    try {
      await apiService.api.post('/smart-preload/configure', config);
    } catch (error) {
      console.error('Error configuring preload service:', error);
      throw error;
    }
  }

  /**
   * Get preload statistics
   */
  async getStats(): Promise<any> {
    try {
      const response = await apiService.api.get('/smart-preload/stats');
      return response.data;
    } catch (error) {
      console.error('Error getting preload stats:', error);
      throw error;
    }
  }

  /**
   * Emergency stop all preload operations
   */
  async emergencyStop(): Promise<void> {
    try {
      await apiService.api.post('/smart-preload/emergency-stop');
    } catch (error) {
      console.error('Error emergency stopping preload:', error);
      throw error;
    }
  }

  /**
   * Start monitoring status updates
   */
  startStatusMonitoring(
    onStatusUpdate: (status: OverallStatus) => void,
    onMemoryUpdate: (memory: MemoryStats) => void,
    intervalMs: number = 2000
  ): void {
    this.onStatusUpdate = onStatusUpdate;
    this.onMemoryUpdate = onMemoryUpdate;

    this.statusUpdateInterval = setInterval(async () => {
      try {
        const [status, memory] = await Promise.all([
          this.getOverallStatus(),
          this.getMemoryStats()
        ]);

        if (this.onStatusUpdate) {
          this.onStatusUpdate(status);
        }

        if (this.onMemoryUpdate) {
          this.onMemoryUpdate(memory);
        }
      } catch (error) {
        console.error('Error in status monitoring:', error);
      }
    }, intervalMs);
  }

  /**
   * Stop monitoring status updates
   */
  stopStatusMonitoring(): void {
    if (this.statusUpdateInterval) {
      clearInterval(this.statusUpdateInterval);
      this.statusUpdateInterval = null;
    }
    this.onStatusUpdate = null;
    this.onMemoryUpdate = null;
  }

  /**
   * Smart preload for slideshow with automatic timing updates
   */
  async smartPreloadForSlideshow(
    playlistId: number,
    currentIndex: number,
    slideDuration: number,
    formats: string[] = ['webp', 'avif']
  ): Promise<void> {
    try {
      // Update timing first
      await this.updateSlideshowTiming(playlistId, currentIndex, slideDuration);

      // Start smart preload
      await this.preloadPlaylist({
        playlist_id: playlistId,
        current_image_index: currentIndex,
        slide_duration: slideDuration,
        formats
      });
    } catch (error) {
      console.error('Error in smart preload for slideshow:', error);
      throw error;
    }
  }

  /**
   * Get preload recommendations for a playlist
   */
  async getPreloadRecommendations(playlistId: number): Promise<{
    recommended: boolean;
    reason: string;
    image_count: number;
    total_size_mb: number;
    estimated_preload_time_seconds: number;
    formats_recommended: string[];
  }> {
    try {
      // This would call the existing preload recommendations endpoint
      const response = await apiService.api.get(`/slideshow-preload/recommendations/${playlistId}`);
      return response.data;
    } catch (error) {
      console.error('Error getting preload recommendations:', error);
      throw error;
    }
  }

  /**
   * Check if preload is recommended for current conditions
   */
  async isPreloadRecommended(playlistId: number): Promise<boolean> {
    try {
      const recommendations = await this.getPreloadRecommendations(playlistId);
      const memoryStats = await this.getMemoryStats();
      
      // Check if memory pressure is low enough
      const memoryOk = memoryStats.used_memory_percent < 70;
      
      return recommendations.recommended && memoryOk;
    } catch (error) {
      console.error('Error checking preload recommendation:', error);
      return false;
    }
  }
}

// Singleton instance
export const smartPreloadService = new SmartPreloadService();














